File: src/domain/summary.js
/**
 * Resumo diário e pareamento de intervalos de sono (puro).
 * Portado de app.js: funções recebem os eventos como argumento (em vez de ler
 * o STORE global), para serem testáveis e portáveis.
 *
 * `babyEvents` = eventos NÃO-deletados e NÃO-compromisso (STORE.babyEvents()).
 */
import { dayKey } from "./time.js";

/**
 * Casa início→fim cronologicamente, mesmo cruzando a meia-noite.
 * O intervalo é atribuído ao dia em que o sono COMEÇOU.
 * (Ordena defensivamente; app.js já passa a lista ordenada por ts.)
 */
export function buildIntervals(events, startType, endType) {
  const sorted = [...events].sort((a, b) => new Date(a.ts) - new Date(b.ts));
  const intervals = [];
  let open = null;
  for (const e of sorted) {
    if (e.type === startType) {
      open = e;
    } else if (e.type === endType && open) {
      intervals.push({
        start: new Date(open.ts),
        end: new Date(e.ts),
        ms: new Date(e.ts) - new Date(open.ts),
        dayKey: dayKey(new Date(open.ts)),
      });
      open = null;
    }
  }
  return intervals;
}

/** Resumo de um dia (`date` no formato YYYY-MM-DD). */
export function summaryForDay(babyEvents, date) {
  const evts = babyEvents.filter((e) => dayKey(new Date(e.ts)) === date);
  const count = (t) => evts.filter((e) => e.type === t).length;

  const milkMl = evts
    .filter((e) => e.type === "milk")
    .reduce((s, e) => s + (Number(e.amountMl) || 0), 0);

  const naps = buildIntervals(babyEvents, "nap_start", "nap_end").filter((i) => i.dayKey === date);
  const napMs = naps.reduce((s, i) => s + i.ms, 0);

  const nights = buildIntervals(babyEvents, "night_start", "wake_morning").filter(
    (i) => i.dayKey === date
  );
  const nightMs = nights.reduce((s, i) => s + i.ms, 0);

  return {
    milkMl,
    milkCount: count("milk"),
    napMs,
    napCount: naps.length,
    nightMs,
    nightCount: nights.length,
    diapers: count("diaper_wet") + count("diaper_poop"),
    poops: count("diaper_poop"),
    meals: count("snack") + count("lunch") + count("dinner"),
    sick: count("sick") > 0,
    total: evts.length,
    events: evts,
  };
}

File: flutter_starter/lib/domain/summary.dart
/// Resumo diário e pareamento de intervalos de sono (portado de `app.js`).
library;

import 'event.dart';

/// Chave local YYYY-MM-DD (respeita o fuso do aparelho, como no PWA).
String dayKey(DateTime d) {
  String two(int n) => n.toString().padLeft(2, '0');
  return '${d.year}-${two(d.month)}-${two(d.day)}';
}

class SleepInterval {
  final DateTime start;
  final DateTime end;
  final int ms;
  final String dayKey; // dia em que o sono COMEÇOU

  const SleepInterval(this.start, this.end, this.ms, this.dayKey);
}

/// Casa início→fim cronologicamente, mesmo cruzando a meia-noite.
/// O intervalo é atribuído ao dia em que o sono começou.
List<SleepInterval> buildIntervals(
  List<Event> events,
  EventType startType,
  EventType endType,
) {
  final list = events.where((e) => !e.deleted).toList()
    ..sort((a, b) => a.ts.compareTo(b.ts));
  final out = <SleepInterval>[];
  Event? open;
  for (final e in list) {
    if (e.type == startType) {
      open = e; // dois inícios seguidos: o mais recente vale
    } else if (e.type == endType && open != null) {
      final ms = e.ts.difference(open.ts).inMilliseconds;
      out.add(SleepInterval(open.ts, e.ts, ms, dayKey(open.ts)));
      open = null;
    }
  }
  return out;
}

class DaySummary {
  final int milkMl;
  final int milkCount;
  final int napMs;
  final int napCount;
  final int nightMs;
  final int nightCount;
  final int diapers;
  final int poops;
  final int meals;

  const DaySummary({
    required this.milkMl,
    required this.milkCount,
    required this.napMs,
    required this.napCount,
    required this.nightMs,
    required this.nightCount,
    required this.diapers,
    required this.poops,
    required this.meals,
  });
}

/// Resumo de um dia (`dateKey` = YYYY-MM-DD).
DaySummary summaryForDay(List<Event> all, String dateKey) {
  final active = all.where((e) => !e.deleted).toList();
  final evts = active.where((e) => dayKey(e.ts) == dateKey).toList();

  int count(EventType t) => evts.where((e) => e.type == t).length;

  final milkMl = evts
      .where((e) => e.type == EventType.milk)
      .fold<int>(0, (s, e) => s + (e.amountMl ?? 0));

  final naps = buildIntervals(active, EventType.napStart, EventType.napEnd)
      .where((i) => i.dayKey == dateKey)
      .toList();
  final nights =
      buildIntervals(active, EventType.nightStart, EventType.wakeMorning)
          .where((i) => i.dayKey == dateKey)
          .toList();

  final napMs = naps.fold<int>(0, (s, i) => s + i.ms);
  final nightMs = nights.fold<int>(0, (s, i) => s + i.ms);

  return DaySummary(
    milkMl: milkMl,
    milkCount: count(EventType.milk),
    napMs: napMs,
    napCount: naps.length,
    nightMs: nightMs,
    nightCount: nights.length,
    diapers: count(EventType.diaperWet) + count(EventType.diaperPoop),
    poops: count(EventType.diaperPoop),
    meals: count(EventType.milk) + count(EventType.snack) + count(EventType.lunch),
  );
}
